### PR TITLE
Add synchronous parameter to MLflowLogger

### LINKED ISSUE #19710: Expose client argument in `MLFlowLogger`
### Description & Motivation

MLflow has an [async autologging client](https://github.com/mlflow/mlflow/blob/2248511d5d86bbd751a0fea4241f78be7bc8f3c8/mlflow/utils/autologging_utils/client.py#L89). It also allows logging metrics asynchronously by passing `synchronous=False` to the various `log_` methods.

It would be useful to support passing in a client argument to the `MLflowLogger` to make it possible to use the async client. Or alternatively, exposing a `synchronous` argument that would be passed down to the `log_` methods.

Maybe this autologging client should be the default? This is already the default when using the `mlflow` autolog: https://github.com/mlflow/mlflow/blob/master/mlflow/pytorch/_lightning_autolog.py#L472.

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

cc @borda

### PR DESCRIPTION
**Pull request by clumsy**
## What does this PR do?

Adding an asynchronous version of `MLflowLogger`. It will do the I/O in a separate thread, thus not blocking the caller (e.g. main training thread).
Note that the values and timestamps of the metrics, params and tags are created synchronously prior the asynchronous I/O thus preserving the original content.

Fixes  #19710

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19639.org.readthedocs.build/en/19639/

### CODE REVIEW

**Comment by awaelchli** [comment]
@clumsy Thanks for the PR. I'm not sure we should do this. I couldn't believe that MLFlowLogger is blocking, so I did a quick search and found this: https://github.com/mlflow/mlflow/issues/1550#issuecomment-1024492066
Maybe it's too old, and things have changed, but it says there by the maintainer that the mlflow client is not thread-save. Furthermore, they suggest using `log_batch`, which we already do in the regular MLFlowLogger. There is a `synchronous=True` flag, maybe we should use that?
https://mlflow.org/docs/latest/python_api/mlflow.client.html?highlight=log_batch#mlflow.client.MlflowClient.log_batch

**Comment by clumsy** [comment]
@awaelchli The issue you linked suggest using the workaround provided in this PR - i.e. don't make asynchronous calls using fluent API like `log_metrics` but use `log_batch` instead. And it's not even that much related to not being thread-safe (if you check latest impl that makes use of a new `synchronous` flag is simply passed to the `log_batch` call [here](https://github.com/mlflow/mlflow/blob/79db4bc0ba7d147f168d5194ec0f91a5f7a8e442/mlflow/tracking/fluent.py#L867), the problem is the data like timestamp is captured within `log_metrics` itself [here](https://github.com/mlflow/mlflow/blob/79db4bc0ba7d147f168d5194ec0f91a5f7a8e442/mlflow/tracking/fluent.py#L863), hence it may not match the time of the original invocation.

The call to `log_batch` (at least with `synchronous=True` which is the default) is blocking (sends requests, awaits responses), dispatching downstream impl from [here](https://github.com/mlflow/mlflow/blob/79db4bc0ba7d147f168d5194ec0f91a5f7a8e442/mlflow/store/tracking/rest_store.py#L340) eventually making the blocking calls [here](https://github.com/mlflow/mlflow/blob/79db4bc0ba7d147f168d5194ec0f91a5f7a8e442/mlflow/utils/rest_utils.py#L203) for REST impl.

Bumping the bounds of `mlflow` dependency also works but I think maybe unnecessary Here's why in my opinion:
- data is safely published, a defensive copy is created before calling `log_batch()`, e.g. by us in this PR and `log_metrics()` alike
- GIL provides a happens-before edge to make make sure the data we captured in the calling thread will be the same in the dispatching thread
- current [log()](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.LightningModule.html#lightning.pytorch.core.LightningModule.log) API does not return anything, so we don't need to block eventually or even return a `Future`
- the new `synchronous` flag is still experimental
- this implementation is basically the same as [AsyncCheckpointIO](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.plugins.io.AsyncCheckpointIO.html) but for metrics/tags instead of the checkpoint
- this impl can be marked experimental and remain opt-in

The difference this tweak brings is very significant in our experiments, basically making the calls to `log()` to publish metrics to MLflow free from the perspective of the training thread which is no longer blocked by the I/O of this `Logger`.

**Comment by awaelchli** [comment]
@clumsy I just vote against exposing a feature flag as an entire subclass of the that logger. That's just not the Lightning way. Let's investigate adding a flag to the existing logger and passing down the synchronous flag to the `log_batch` method. It's fine if it's experimental, we wouldn't enable it by default.

**Comment by clumsy** [comment]
A small correction, @awaelchli - this PR is _not_ using the experimental feature flag from the recent MLflow versions.
In fact it's not using MLflow's asynchronous implementation altogether and makes asynchronous metric, tag and parameter publishing available for _any_ version of MLflow.

**Comment by awaelchli** [comment]
Sure, I'm fine with this being an implementation detail. I still don't see enough justification to make this an entirely new class. I stay with my recommendation: `MLFlowLogger(async=True|False)` with False being the default.

**Comment by dannyfriar** [comment]
> How about this then, @awaelchli:
> 1. We add `async` flag to `MLFlowLogger`
> 1. We check if the user passed `async=True`
> 1. If Yes - we check what is `mlflow` version
> 1. If >=2.8.0 we use the new `synchronous=False`
> 1. If <2.8.0 we use a thread pool approach (this PR)

Personally I think that adds a lot of code complexity to do in two different ways. I'd vote for expose a flag and then just log a warning in the case that the mlflow version is below 2.8.0.

I'd also vote for calling the argument `synchronous` to match the mlflow argument.

**Comment by clumsy** [comment]
As for code complexity - isn't it implementation detail? The user never sees this.

Logging, or silently continuing with the synchronous workflow when indicated otherwise by the user violates the principle of least astonishment I'm afraid. I'd rather we assert the flag (or its value `False`) is not used for mlflow<2.8.0

I like the naming suggestion to keep it consistent with mlflow.

**Comment by awaelchli** [comment]
I'm fine with either approach. I would only add the manual implementation if upgrading mlflow adds barriers for users. But no strong opinion on this.

**Comment by dannyfriar** [comment]
It is implementation detail but still adds more code to the project to maintain for relatively little benefit IMO (supporting async in old mlflow versions) . The addition of the `synchronous`flag alone is a tiny change. 

I think raising an exception in that case rather than logging a warning makes sense.

**Comment by clumsy** [comment]
Please check the new version of this PR, @dannyfriar, @awaelchli 
Added optional `synchronous` flag to `MLFlowLogger`) that defaults to `None` - which leaves it for `mlflow` discretion to decide, e.g. if they decide to change the default behavior in the future.
When the flag is provided we will pass it via kwargs where applicable (for us it's just `log_batch` currently), but we verify the current `mlflow` version supports it.
Unfortunately this means that the users of older versions of `mlflow` (e.g. because their MLflow endpoint is <2.8.0 and they cannot use newer clients) won't get the luxury of asynchronous MLflow logger.

**Comment by dannyfriar** [comment]
Nice - that LGTM!

File: src/lightning/pytorch/loggers/mlflow.py
Comment (awaelchli) [inline src/lightning/pytorch/loggers/mlflow.py:146]: I suggest to make this name specific for how it's used.
```suggestion
        self._log_batch_kwargs = {} if synchronous is None else {"synchronous": synchronous}
```

File: src/lightning/pytorch/loggers/mlflow.py
Comment (awaelchli) [inline src/lightning/pytorch/loggers/mlflow.py:104]: I suggest to explain in plain words what it does. This way, it's easier to discover for users who may not also look at the mlflow docs at the same time.

```suggestion
        synchronous: Whether to block the execution every logging call until complete. Defaults to True (blocking) if unspecified.
            Setting to False requires mlflow >= 2.8.0.
```

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/loggers/mlflow.py
Comment (clumsy) [reply src/lightning/pytorch/loggers/mlflow.py:104]: @awaelchli I change for sure, but we can't guarantee this will be how it works - we are leaving this up to `mlflow` that can simply ignore this flag (potentially). The default is actually - up to `mlflow` too.
Had we used our own implementation with the thread pool we could have guaranteed this.

How about `Hints mlflow whether to block the execution for every logging call until complete where applicable.`? I'd like to keep `where applicable` since `log_artifacts` does not have this options, it's always synchronous.

Comment (awaelchli) [reply src/lightning/pytorch/loggers/mlflow.py]: Yes sure, feel free to add that part back. My main ask is just to explain it plainly and briefly so it's clear.

**Review by carmocca — APPROVED** [review]

**Comment by awaelchli** [comment]
Thank you for the feature @clumsy and @dannyfriar for discussion!

**Pull request merged**
